lib/pbench/test/unit/server/test_cache_manager.py

```diff
@@ -244,17 +244,55 @@ def test_duplicate(
         monkeypatch.setattr(Tarball, "_get_metadata", fake_get_metadata)
         cm = CacheManager(server_config, make_logger)
 
-        # Create a tarball file in the expected destination directory
+        # Create a tarball file in the expected destination directory: the
+        # subsequent create should report a duplicate.
         controller = cm.archive_root / "ABC"
         controller.mkdir()
         (controller / source_tarball.name).write_text("Send in the clones")
-
-        # Attempting to create a dataset from the md5 file should result in
-        # a duplicate dataset error
         with pytest.raises(DuplicateTarball) as exc:
             cm.create(source_tarball)
         assert exc.value.tarball == Dataset.stem(source_tarball)
 
+    @pytest.mark.parametrize("allow", (".md5", ""))
+    def test_move_fails(
+        self, monkeypatch, selinux_disabled, server_config, make_logger, tarball, allow
+    ):
+        src: list[Path] = []
+        dest: list[Path] = []
+        real_move = shutil.move
+
+        def mymove(source: Path, destination: Path, *args, **kwargs) -> Path:
+            src.append(source)
+            if destination.is_dir():
+                d = destination / source.name
+            else:
+                d = destination
+            dest.append(d)
+            if source.suffix == allow:
+                return real_move(source, destination, *args, **kwargs)
+            raise Exception("I'm broken")
+
+        ulink: list[Path] = []
+        ok: list[bool] = []
+
+        def unlink(self, missing_ok: bool = False):
+            ulink.append(self)
+            ok.append(missing_ok)
+
+        source_tarball, source_md5, md5 = tarball
+        monkeypatch.setattr(Tarball, "_get_metadata", fake_get_metadata)
+        cm = CacheManager(server_config, make_logger)
+        real_move = shutil.move
+        monkeypatch.setattr("pbench.server.cache_manager.shutil.move", mymove)
+        monkeypatch.setattr(Path, "unlink", unlink)
+        with pytest.raises(Exception) as e:
+            cm.create(source_tarball)
+        assert str(e.value) == "I'm broken"
+        assert src == [source_md5] + ([source_tarball] if allow else [])
+        for i in range(len(ulink)):
+            assert ulink[i] == dest[i]
```

Comment:
It comes down to how well we document the parameters of an unexpected problem if something changes radically in the code without an equivalent change here. I decided to keep it simple and test what's supposed to happen without getting too fancy about it.

Reply:
Well, the unit tests are supposed to function (at least in part) as an executable specification for the behavior of the CUT.  We should not (necessarily) have to modify the unit tests every time we modify the code, and, in fact, doing so undermines the usefulness of the tests.

In this case, since `src`, `dest`, `unlink`, and `ok` are all _supposed_ to have the same number of entries (when the CUT is operating as expected), we should be able to use any of them to drive this loop.  Since `src` drives the test overall, it would be the best one to use here; whereas, `ulink` is the most likely be disturbed by changes to either the test or the CUT, so it's probably the worst choice, unless it is buttressed by an assertion which makes sure that its length matches that of `dest`.

So, it's not about being fancy -- it's about making sure that the test checks what you intend for it to check and that it doesn't produce a false-PASS if the CUT (or the test) fails in just the right way.